File: lib-core/lib-util/src/org/aiotrade/lib/util/swing/table/AttributiveCellRenderer.java
/*
 * (swing1.1beta3)
 */
package org.aiotrade.lib.util.swing.table;

import java.awt.Color;
import java.awt.Component;
import java.awt.Font;
import javax.swing.JTable;
import javax.swing.UIManager;
import javax.swing.border.Border;
import javax.swing.table.DefaultTableCellRenderer;
import javax.swing.table.TableModel;

/**
 * @version 1.0 11/22/98
 */
public class AttributiveCellRenderer extends DefaultTableCellRenderer {

    public AttributiveCellRenderer() {
        super();
    }

    @Override
    public Component getTableCellRendererComponent(JTable table, Object value,
            boolean isSelected, boolean hasFocus, int row, int column) {
        Color foreground = null;
        Color background = null;
        Font font = null;
        TableModel model = table.getModel();
        if (model instanceof AttributiveCellTableModel) {
            CellAttribute cellAttr = ((AttributiveCellTableModel) model).getCellAttribute();
            if (cellAttr instanceof ColoredCell) {
                foreground = ((ColoredCell) cellAttr).getForeground(row, column);
                background = ((ColoredCell) cellAttr).getBackground(row, column);
                setHorizontalAlignment(((ColoredCell) cellAttr).getHorizontalAlignment(row, column));
            }
            if (cellAttr instanceof CellFont) {
                font = ((CellFont) cellAttr).getFont(row, column);
            }
        }

        if (isSelected) {
            setForeground(foreground != null ? foreground : table.getSelectionForeground());
            setBackground(table.getSelectionBackground());
        } else {
            setForeground(foreground != null ? foreground : table.getForeground());
            setBackground(background != null ? background : table.getBackground());
        }

        setFont(font != null ? font : table.getFont());

        if (hasFocus) {
            Border border = null;
            if (isSelected) {
                border = UIManager.getBorder("Table.focusSelectedCellHighlightBorder");
            }
            if (border == null) {
                border = UIManager.getBorder("Table.focusCellHighlightBorder");
            }
            setBorder(border);

            if (!isSelected && table.isCellEditable(row, column)) {
                setForeground((foreground != null) ? foreground : UIManager.getColor("Table.focusCellForeground"));
                setBackground(UIManager.getColor("Table.focusCellBackground"));
            }
        } else {
            setBorder(noFocusBorder);
        }

        setValue(value);

        return this;
    }

    @Override
    protected void setValue(Object value) {
        setText(value == null ? "" : value.toString());
    }
}

File: libs/lib.util/src/main/scala/org/aiotrade/lib/util/swing/table/AttributiveCellRenderer.scala
package org.aiotrade.lib.util.swing.table

import java.awt.Color
import java.awt.Component
import java.awt.Font
import javax.swing.JTable
import javax.swing.UIManager
import javax.swing.border.{Border, EmptyBorder}
import javax.swing.table.DefaultTableCellRenderer

/**
 * @version 1.0 11/22/98
 */
object AttributiveCellRenderer {
  private val noFocusBorder: Border = new EmptyBorder(1, 1, 1, 1)
}
class AttributiveCellRenderer extends DefaultTableCellRenderer {

  override def getTableCellRendererComponent(table: JTable, value: Object,
                                             isSelected: Boolean, hasFocus: Boolean,
                                             row: Int, column: Int): Component = {
    var foreground: Color = null
    var background: Color = null
    var font: Font = null
    val model = table.getModel
    if (model.isInstanceOf[AttributiveCellTableModel]) {
      val cellAttr = model.asInstanceOf[AttributiveCellTableModel].getCellAttribute
      if (cellAttr.isInstanceOf[ColoredCell]) {
        foreground = cellAttr.asInstanceOf[ColoredCell].getForeground(row, column)
        background = cellAttr.asInstanceOf[ColoredCell].getBackground(row, column)
        setHorizontalAlignment(cellAttr.asInstanceOf[ColoredCell].getHorizontalAlignment(row, column))
      }
      if (cellAttr.isInstanceOf[CellFont]) {
        font = cellAttr.asInstanceOf[CellFont].getFont(row, column)
      }
    }

    if (isSelected) {
      setForeground(if (foreground != null) foreground else table.getSelectionForeground)
      setBackground(table.getSelectionBackground)
    } else {
      setForeground(if (foreground != null) foreground else table.getForeground)
      setBackground(if (background != null) background else table.getBackground)
    }

    setFont(if (font != null) font else table.getFont)

    if (hasFocus) {
      var border: Border = if (isSelected) {
        UIManager.getBorder("Table.focusSelectedCellHighlightBorder")
      } else {
        null
      }
      if (border == null) {
        border = UIManager.getBorder("Table.focusCellHighlightBorder")
      }
      setBorder(border)

      if (!isSelected && table.isCellEditable(row, column)) {
        setForeground(if (foreground != null) foreground else UIManager.getColor("Table.focusCellForeground"))
        setBackground(UIManager.getColor("Table.focusCellBackground"))
      }
    } else {
      setBorder(AttributiveCellRenderer.noFocusBorder)
    }

    setValue(value)

    this
  }

  override protected def setValue(value: Object): Unit = {
    setText(if (value == null) "" else value.toString)
  }
}


